packages/form-js-viewer/src/render/components/form-fields/Taglist.js

```diff
@@ -50,43 +49,43 @@ export default function Taglist(props) {
     values: options
   } = useValuesAsync(field);
 
+  // We cache a map of option values to their index so that we don't need to search the whole options array every time to correlate the label
+  const valueToOptionMap = useMemo(() => Object.assign({}, ...options.map((o, x) => ({ [o.value]:  options[x] }))), [ options ]);
 
   // Usage of stringify is necessary here because we want this effect to only trigger when there is a value change to the array
   useEffect(() => {
     if (loadState === LOAD_STATES.LOADED) {
-      const selectedValues = values.map(v => options.find(o => o.value === v)).filter(v => v !== undefined);
-      setSelectedValues(selectedValues);
+      setFilteredOptions(options.filter((o) => o.label && (o.label.toLowerCase().includes(filter.toLowerCase())) && !values.includes(o.value)));
     }
     else {
-      setSelectedValues([]);
+      setFilteredOptions([]);
     }
-  }, [ JSON.stringify(values), options, loadState ]);
+  }, [ filter, JSON.stringify(values), options, loadState ]);
 
   useEffect(() => {
-    if (loadState === LOAD_STATES.LOADED) {
-      setFilteredValues(options.filter((o) => o.label && (o.label.toLowerCase().includes(filter.toLowerCase())) && !values.includes(o.value)));
-    }
-    else {
-      setFilteredValues([]);
-    }
-  }, [ filter, JSON.stringify(values), options ]);
-
-  useEffect(() => {
-    setHasValuesLeft(selectedValues.length < options.length);
-  }, [ selectedValues.length, options.length ]);
+    setHasOptionsLeft(options.length > values.length);
+  }, [ options.length, values.length ]);
 
   const onFilterChange = ({ target }) => {
     setIsEscapeClose(false);
     setFilter(target.value);
   };
 
-  const selectValue = (option) => {
-    setFilter('');
-    props.onChange({ value: [ ...values, option.value ], field });
+  const selectValue = (value) => {
+    if (filter) {
+      setFilter('');
+    }
+
+    // Ensure values cannot be double selected due to latency
+    if (values.at(-1) === value) {
```

Comment:
I can't find any test case for this?

Reply:
Could we realistically rely on test cases for this? This is after all an issue that only occurs due to large re-renders. Could we replicate the lagging conditions under test? The bug was not consistent in the first place so I see a test like this could very likely stay green even if the issue is back.

Reply:
I'm planning on creating a custom run script to stress test our components that are supposed to handle large datasets, tracked here: https://github.com/bpmn-io/form-js/issues/288. Perhaps this could be something better to handle these kinds of cases. It's not automatic testing but it would help.

Reply:
If we can't reliably reproduce it by setting such a value in the existing TagList specs, we could consider to extract the `selectValue` functionality, export it and have a dedicated unit test in the TagList specs.

```js
// internal implementation of the component
const _selectValue = (value) => {
 return selectValue(value, values, setFilter, props.onChange);
}

// helper, which we test
export function selectValue(value, values, setFilter, onChange) {
  // ...
}
```

We should consider it only if it makes sense and it's worth the effort. My gut feeling just notified me because we fixed something in the PR and didn't provide a test case for it. 😅

Reply:
@pinussilvestrus Where would you have the helper function? I actually think this would be the most robust thing to do, albeit a bit intrusive. I guess the question is, do we want to complicate things to this extent to test the rather simple condition? In a way, it does also disrupt the code. 

I personally think it's too disruptive.

Reply:
@pinussilvestrus Where would you have the helper function? I actually think this would be the most robust thing to do, albeit a bit intrusive. I guess the question is, do we want to complicate things to this extent to test the rather simple condition? In a way, it does also disrupt the code. 

I personally think it's too disruptive.

Reply:
Agree, It's ok to keep things simple (as they are) in this case 👍